agents/agents/guard/fraud.go

```diff
@@ -0,0 +1,792 @@
+package guard
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/synapsecns/sanguine/agents/types"
+	"github.com/synapsecns/sanguine/core/retry"
+)
+
+// handleSnapshot checks a snapshot for invalid states.
+// If an invalid state is found, initiate slashing and submit a state report.
+//
+//nolint:cyclop,gocognit
+func (g Guard) handleSnapshot(ctx context.Context, log ethTypes.Log) error {
+	fraudSnapshot, err := g.inboxParser.ParseSnapshotAccepted(log)
+	if err != nil {
+		return fmt.Errorf("could not parse snapshot accepted: %w", err)
+	}
+
+	// Verify each state in the snapshot.
+	for si, s := range fraudSnapshot.Snapshot.States() {
+		stateIndex, state := si, s
+		isSlashable, err := g.isStateSlashable(ctx, state)
+		if err != nil {
+			return fmt.Errorf("could not handle state: %w", err)
+		}
+		if !isSlashable {
+			continue
+		}
+
+		// Initiate slashing on origin.
+		_, err = g.txSubmitter.SubmitTransaction(ctx, big.NewInt(int64(state.Origin())), func(transactor *bind.TransactOpts) (tx *ethTypes.Transaction, err error) {
+			tx, err = g.domains[state.Origin()].LightInbox().VerifyStateWithSnapshot(
+				transactor,
+				int64(stateIndex),
+				fraudSnapshot.Payload,
+				fraudSnapshot.Signature,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("could not verify state with snapshot: %w", err)
+			}
+
+			return
+		})
+		if err != nil {
+			return fmt.Errorf("could not submit VerifyStateWithSnapshot tx: %w", err)
+		}
+
+		// Check if we should submit the state report.
+		shouldSubmit, err := g.shouldSubmitStateReport(ctx, fraudSnapshot)
```

Comment:
I feel like should submit logic should be applied to both Synapse Chain and remote chain. For instance:
- Someone has reported a fraudulent Notary on SynChain, but not on remote chain. Report on remote chain only.
- Someone has reported a fraudulent Notary on remote chain, but not on SynChain. Report on SynChain only.
- No one has reported a fraudulent Notary. Report on both chains.